records/2023-03-28.secret-management.md

```diff
@@ -0,0 +1,189 @@
+
+A positive of vault is that it is very flexible in the authentication methods that can be used. The default vault root administrator will use temporary token authentication, but this is not the recommended approach for day to day use. Instead, authentication can be done via standard username/password, OIDC login to an external IDP (GitHub etc), AWS account delegation, Kubernetes, and more. This makes it very easy to integrate virtually any system within vault. Vault uses the concept of roles that are assigned to an authentication method. The role is a group of policies and permissions that can be assigned to a user who authenticates with this auth method. There will be a default role that will be assigned to the auth method, but additional roles can be specified and requested during login.
+
+Vault uses policies for permission management. A policy is a JSON structured document that defines the permissions to the individual secret engines. Policies use keywords such as "Read","Write" and "List" to define how an entity is permitted to interact with a specific secret engine. These policies can be signed directly to an entity, or assigned to a group (which is a group of entities), or assigned to a role (which is a collection of policies and groups). 
+
+### Vault - Self Hosted (Kubernetes)
+
+#### Cost
+
+Self hosting vault is free from a licensing perspective. There will be a minor additional cost on compute resources within the Kubernetes cluster, however Vault is not typically resource intensive from a compute perspective.
+
+#### Administration Overhead
+
+Vault can be installed in Kubernetes using the Vault Helm chart. This very quickly and easily sets up and configures vault within the Kubernetes cluster. The issue with self-hosted vault is always with Day 2 operations. As a best practice, in a production environment, Vault should be configured in High Availability mode which requires additional configuration. Since the secret engine data resides within the Kubernetes cluster, an automated backup tool such as Velero will be required to ensure that data is protected. If a vault pod is moved from one K8s node to another, the vault will automatically seal itself requiring either manual intervention to unseal, or the implementation of the vault auto-unseal operator. All of this is on top of the day to day IAM administration. Because of this, self hosted vault should only be considered if there are requirements that can not be met via HCP Vault (data residency, private links, cost etc).
```

Comment:
Good to mention private links and data residency, as this option would actually be an advantage if we have restrictions in place that prevent us from going forward with a SaaS option. It would be worth to have a dedicated section on policies (data residency), technical restrictions, potential red flags.

Reply:
Sure I'll add a section on that!